x/upgrade/abci.go

```diff
@@ -9,17 +9,29 @@ import (
 )
 
 // BeginBlock will check if there is a scheduled plan and if it is ready to be executed.
+// If skip upgrade flag is set, it will skip and clear the upgrade plan
 // If it is ready, it will execute it if the handler is installed, and panic/abort otherwise.
 // If the plan is not ready, it will ensure the handler is not registered too early (and abort otherwise).
 //
-// The prupose is to ensure the binary is switch EXACTLY at the desired block, and to allow
+// The purpose is to ensure the binary is switched EXACTLY at the desired block, and to allow
 // a migration to be executed if needed upon this switch (migration defined in the new binary)
-func BeginBlocker(k Keeper, ctx sdk.Context, _ abci.RequestBeginBlock) {
+func BeginBlocker(k Keeper, ctx sdk.Context, _ abci.RequestBeginBlock, skipUpgrade bool) {
+
 	plan, found := k.GetUpgradePlan(ctx)
 	if !found {
 		return
 	}
+
 	if plan.ShouldExecute(ctx) {
+		//To make sure clear upgrade is executed at the same block
+		if skipUpgrade {
+			// If skip upgrade has been set, we clear the upgrade plan
+			skipUpgradeMsg := fmt.Sprintf("UPGRADE \"%s\" SKIPPED at %s: %s", plan.Name, plan.DueAt(), plan.Info)
+			ctx.Logger().Info(skipUpgradeMsg)
+			k.ClearUpgradePlan(ctx)
+			return
+		}
+
```

Comment:
@aaronc yes that is true. Instead of taking height and cross checking, I propose to clear the flag (unset) once it is used. Because if we are starting the node with skip-upgrade flag, we intend to skip that particular upgrade and the flag should not carry any value after that process.

Reply:
Anil answer seems reasonable, assuming people manually set this flag. If they start putting it in the systems restart script, then it is problematic. But that is a bigger problem.

I see Aaron's approach as safer, but also harder to use. 

We should consider this will be used in emergency situations and there will be a long period before a valid upgrade proposal. I am not sure which is better, but having user stories in mind can help design.

As an alternative, I could suggest we include the name of the plan instead of height but use it the same way

As another alternative, it could take an optional height field, which will enforce Aaron's logic present, and default to Anils proposal (perform once and unset flag) if left as default (0). Here I would use 0 as skip any plan, positive as only skip at height x, and negative as flag used, ignore it.

Reply:
Here are some user stories:

* the most common - an upgrade is supposed to happen, but an error occurs during the upgrade handler and thus can't proceed, the validators agree to use `--unsafe-abort-upgrade` to keep going - they do know the height at which this happens
* validators agree in advance to skip an upgrade outside of governance - they may not know the height, BUT they can still want until the upgrade time hits and disable automatic upgrades, and then they can use the height to continue
* someone is syncing a full node from genesis - they need to know which upgrades to skip because these happen outside of the block history. In this case, they need to know the heights because the same named upgrade could have occurred later after a failure - i.e. say `upgrade1` was scheduled for height 5000 and failed, later `upgrade1` got rescheduled for 10000 and succeeds, so we only want `--unsafe-skip-upgrade 5000` not `--unsafe-skip-upgrade upgrade1` for full nodes so they can sync from genesis

So I know it is a bit more complex, but from this the solution that covers the most use cases and also seems the safest/most correct is we can allow multiple `--unsafe-skip-upgrade <height>` flags to handle potentially multiple aborted upgrades in a chain history - in particular this helps with syncing a full node from genesis which is a fairly common use case.

Thoughts?

Reply:
I agree with @aaronc, this works best considering all the user stories. Instead of handling each case differently, `--unsafe-skip-upgrade <height>` can work commonly to handle all the cases

Reply:
> * validators agree in advance to skip an upgrade outside of governance - they may not know the height, BUT they can still want until the upgrade time hits and disable automatic upgrades, and then they can use the height to continue

Regarding this, is there a way to disable automatic upgrades now?

Reply:
Well automatic upgrades can be disabled by either 1) not using `cosmosd` at all or 2) not providing `cosmosd` with an upgrade binary and not setting `DAEMON_ALLOW_DOWNLOAD_BINARIES`. So if you're either not using cosmosd or cosmosd has nothing to upgrade to, the binary should just panic at that height and the operator needs to intervene.